File: iOS/Igatha/Services/DisasterDetector.swift
//
//  DisasterDetector.swift
//  Igatha
//
//  Created by Nizar Mahmoud on 12/10/2024.
//

import Foundation

class DisasterDetector: SensorDelegate {
    weak var delegate: DisasterDetectorDelegate?
    
    private let eventTimeWindow: TimeInterval
    private var eventTimes: [SensorType: Date] = [:]
    
    private let accelerometerSensor: AccelerometerSensor
    private let gyroscopeSensor: GyroscopeSensor
    private let barometerSensor: BarometerSensor
    
    public var isAvailable: Bool {
        return (
            accelerometerSensor.isAvailable &&
            gyroscopeSensor.isAvailable &&
            barometerSensor.isAvailable
        )
    }
    
    init(
        accelerationThreshold: Double,
        rotationThreshold: Double,
        pressureThreshold: Double,
        eventTimeWindow: TimeInterval
    ) {
        self.eventTimeWindow = eventTimeWindow
        
        accelerometerSensor = AccelerometerSensor(
            threshold: accelerationThreshold,
            updateInterval: 0.1
        )
        gyroscopeSensor = GyroscopeSensor(
            threshold: rotationThreshold,
            updateInterval: 0.1
        )
        barometerSensor = BarometerSensor(
            threshold: pressureThreshold
        )
        
        accelerometerSensor.delegate = self
        gyroscopeSensor.delegate = self
        barometerSensor.delegate = self
        
        delegate?.disasterDetectorAvailabilityUpdate(isAvailable)
    }
    
    deinit {
        stopDetection()
    }
    
    func startDetection() {
        guard isAvailable else { return }
        
        accelerometerSensor.startUpdates()
        gyroscopeSensor.startUpdates()
        barometerSensor.startUpdates()
        
        delegate?.disasterDetectionStarted()
    }
    
    func stopDetection() {
        accelerometerSensor.stopUpdates()
        gyroscopeSensor.stopUpdates()
        barometerSensor.stopUpdates()
        
        delegate?.disasterDetectionStopped()
    }
    
    // called when a sensor exceeds a threshold
    func sensorExceededThreshold(
        sensorType: SensorType,
        eventTime: Date
    ) {
        eventTimes[sensorType] = eventTime
        
        checkForIncident()
    }
    
    // called when an incident is suspected
    private func checkForIncident() {
        let currentTime = Date()
        
        // ensure all events have occurred
        guard eventTimes.count == 3 else { return }
        
        // ensure all events occurred within the time window
        for (_, eventTime) in eventTimes {
            if currentTime.timeIntervalSince(eventTime) > eventTimeWindow {
                // event is too old; do not consider it
                return
            }
        }
        
        // incident detected
        delegate?.disasterDetected()
    }
}

protocol DisasterDetectorDelegate: AnyObject {
    func disasterDetected()
    
    func disasterDetectionStarted()
    func disasterDetectionStopped()
    
    func disasterDetectorAvailabilityUpdate(_ isAvailable: Bool)
}

File: Android/app/src/main/java/com/nizarmah/igatha/service/DisasterDetector.kt
package com.nizarmah.igatha.service

import android.content.Context
import com.nizarmah.igatha.Constants
import com.nizarmah.igatha.sensor.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import com.nizarmah.igatha.sensor.Sensor as InternalSensor

class DisasterDetector(
    context: Context,
    accelerationThreshold: Double,
    rotationThreshold: Double,
    pressureThreshold: Double,
    private val eventTimeWindow: Long // in milliseconds
) {

    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    private val eventTimes = mutableMapOf<SensorType, Long>()

    private val accelerometerSensor = AcceleratorSensor(
        context,
        threshold = accelerationThreshold,
        updateInterval = Constants.SENSOR_UPDATE_INTERVAL
    )
    private val gyroscopeSensor = GyroscopeSensor(
        context,
        threshold = rotationThreshold,
        updateInterval = Constants.SENSOR_UPDATE_INTERVAL
    )
    private val barometerSensor = BarometerSensor(
        context,
        threshold = pressureThreshold,
        updateInterval = Constants.SENSOR_UPDATE_INTERVAL
    )

    private val _isAvailable = MutableStateFlow(true)
    val isAvailable: StateFlow<Boolean> = _isAvailable.asStateFlow()

    private val _isActive = MutableStateFlow(false)
    val isActive: StateFlow<Boolean> = _isActive.asStateFlow()

    fun startDetection() {
        scope.launch {
            collectSensorEvents(accelerometerSensor)
        }
        scope.launch {
            collectSensorEvents(gyroscopeSensor)
        }
        scope.launch {
            collectSensorEvents(barometerSensor)
        }

        _isActive.value = true
    }

    fun stopDetection() {
        _isActive.value = false

        accelerometerSensor.stopUpdates()
        gyroscopeSensor.stopUpdates()
        barometerSensor.stopUpdates()
        scope.coroutineContext.cancelChildren()
    }

    private suspend fun collectSensorEvents(sensor: InternalSensor) {
        sensor.startUpdates()
        sensor.events.collect { event ->
            eventTimes[event.sensorType] = System.currentTimeMillis()
            checkForIncident()
        }
    }

    private fun checkForIncident() {
        val currentTime = System.currentTimeMillis()
        if (eventTimes.size < 3) return

        if (eventTimes.values.all { currentTime - it <= eventTimeWindow }) {
            // Disaster detected
            scope.launch {
                DisasterEventBus.emitDisasterDetected()
            }
            eventTimes.clear()
        }
    }

    fun deinit() {
        stopDetection()
        scope.cancel()
    }
}
